runtime/src/conn.rs

```diff
@@ -0,0 +1,185 @@
+    url: String,
+    duration: Duration,
+) -> Result<WsClient, Error> {
+    info!("Connecting to the btc-parachain...");
+    timeout(duration, async move {
+        loop {
+            match new_websocket_client(url.clone()).await {
+                Err(Error::JsonRpseeError(JsonRpseeError::TransportError(err))) => {
+                    trace!("could not connect to parachain: {}", err);
+                    delay_for(RETRY_DURATION).await;
+                    continue;
+                }
+                Ok(rpc) => {
+                    info!("Connected!");
+                    return Ok(rpc);
+                }
+                Err(err) => return Err(err),
+            }
+        }
+    })
+    .await?
+}
+
+pub enum RetryPolicy {
+    Never,
+    Always,
+}
+
+impl FromStr for RetryPolicy {
+    type Err = String;
+    fn from_str(code: &str) -> Result<Self, Self::Err> {
+        match code {
+            "never" => Ok(RetryPolicy::Never),
+            "always" => Ok(RetryPolicy::Always),
+            _ => Err("Could not parse input as RetryPolicy".to_string()),
+        }
+    }
+}
+
+pub struct ManagerConfig {
+    pub retry_timeout: Duration,
+    pub retry_policy: RetryPolicy,
+}
+
+pub struct Manager<C: Clone, P: Provider, S: Service<C, P>> {
+    url: String,
+    signer: PolkaBtcSigner,
+    service_config: C,
+    manager_config: ManagerConfig,
+    handle: Handle,
+    _marker: PhantomData<(P, S)>,
+}
+
+impl<C: Clone + Send + 'static, P: Provider + Send, S: Service<C, P>> Manager<C, P, S> {
+    pub fn new(
+        url: String,
+        signer: PolkaBtcSigner,
+        service_config: C,
+        manager_config: ManagerConfig,
+        handle: Handle,
+    ) -> Self {
+        Self {
+            url,
+            signer,
+            service_config,
+            manager_config,
+            handle,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub async fn start(&self) -> Result<(), Error> {
+        loop {
+            let ws_client = new_websocket_client_with_retry(
+                self.url.clone(),
+                self.manager_config.retry_timeout,
+            )
+            .await?;
+
+            let signer = self.signer.clone();
+            let config = self.service_config.clone();
+            let handle = self.handle.clone();
+
+            let (shutdown_sender, shutdown_receiver) = tokio::sync::watch::channel(None);
+            let connected_listener = is_connected(ws_client.clone());
+
+            let _ = tokio::join! {
+                async move {
+                    // TODO: propogate shutdown signal from children
+                    connected_listener.await;
+                    let _ = shutdown_sender.broadcast(Some(()));
+                },
+                async move {
+                    let provider = P::connect(ws_client, signer).await?;
+                    let _ = S::connect(provider, config, handle, shutdown_receiver).await;
```

Comment:
Maybe `connect` should be renamed - this will start the service, and block until it finishes, right? From the name I'd expect this call to return after a connection has been established